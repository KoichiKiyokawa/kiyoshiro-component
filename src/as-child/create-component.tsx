import React, { cloneElement, forwardRef, isValidElement } from "react"

type Options<E extends keyof React.JSX.IntrinsicElements> = {
	baseElement: React.ReactElement
	props: React.ComponentPropsWithoutRef<E>
	preChildren?: React.ReactElement
	postChildren?: React.ReactElement
	outer?: React.FC<React.PropsWithChildren>
	inner?: React.FC<React.PropsWithChildren>
}

type ComponentProps<E extends keyof React.JSX.IntrinsicElements, P> = P &
	(
		| { asChild: true; children: React.ReactElement }
		| ({ asChild?: false } & React.ComponentPropsWithRef<E>)
	)

export function createComponent<E extends keyof React.JSX.IntrinsicElements, P>(
	builder: (
		props: P & React.ComponentPropsWithoutRef<E>,
		info: {
			ref: React.ComponentPropsWithRef<E>["ref"]
		} & (
			| { asChild: true; children: React.ReactElement }
			| { asChild: false; children: React.ReactNode }
		),
	) => Options<E>,
) {
	return forwardRef(
		(
			{ asChild = false, children, ...restProps }: ComponentProps<E, P>,
			ref: React.ComponentPropsWithRef<E>["ref"],
		) => {
			const { baseElement, props, outer, inner, preChildren, postChildren } =
				// @ts-ignore
				builder(restProps, { asChild, children, ref })

			const mainChildElements = (
				<>
					{preChildren}
					{asChild && isValidElement(children) ? children.props.children : children}
					{postChildren}
				</>
			)

			const clonedElement = cloneElement(
				baseElement,
				props,
				inner ? inner({ children: mainChildElements }) : mainChildElements,
			)

			return outer ? outer({ children: clonedElement }) : clonedElement
		},
	) as React.FC<ComponentProps<E, P>>
}
