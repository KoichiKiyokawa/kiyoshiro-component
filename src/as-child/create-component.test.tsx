import { render } from "@testing-library/react"
import { describe, expect, it } from "vitest"
import { createComponent } from "./create-component"

describe("create normal button component", () => {
	const Button = createComponent<
		"button",
		{ variant: "primary" | "secondary"; disabled?: boolean }
	>(
		(
			{ variant, disabled = false, ...buttonProps },
			{ asChild, children, ref },
		) => ({
			baseElement: asChild ? (
				disabled ? (
					<div />
				) : (
					children
				)
			) : (
				<button type="button" ref={ref} disabled={disabled} {...buttonProps} />
			),
			
			props: {
				"data-variant": variant,
				className: "button",
				...(disabled && { "aria-disabled": true }),
			},
		}),
	)

	it.each([
		// as button
		{ name: "primary button", ui: <Button variant="primary">Click me</Button> },
		{
			name: "secondary button",
			ui: <Button variant="secondary">Click me</Button>,
		},
		{
			name: "disabled button",
			ui: (
				<Button variant="primary" disabled>
					Click me
				</Button>
			),
		},

		// as link
		{
			name: "button as link",
			ui: (
				<Button asChild variant="primary">
					<a href="...">link</a>
				</Button>
			),
		},
		{
			name: "button as disabled link",
			ui: (
				<Button asChild variant="primary" disabled>
					<a href="...">link</a>
				</Button>
			),
		},
	])("$name", ({ ui }) => {
		const { asFragment } = render(ui)

		expect(asFragment()).toMatchSnapshot()
	})
})

describe("create button component with decoration props(outer, inner, ...)", () => {
	const Button = createComponent<
		"button",
		{ variant: "primary" | "secondary"; disabled?: boolean }
	>(
		(
			{ variant, disabled = false, ...buttonProps },
			{ asChild, children, ref },
		) => ({
			baseElement: asChild ? (
				disabled ? (
					<div />
				) : (
					children
				)
			) : (
				<button type="button" ref={ref} disabled={disabled} {...buttonProps} />
			),
			props: {
				"data-variant": variant,
				className: "button",
				...(disabled && { "aria-disabled": true }),
			},
			outer: ({children}) => <div className="outer">{children}</div>,
			inner: ({children}) => <div className="inner">{children}</div>,
			preChildren: <><span className="pre1">pre1</span><span className="pre2">pre2</span></>,
		 	postChildren: <><span className="post1">post1</span><span className="post2">post2</span></>,
		}),
	)

	it.each([
		{ name: "primary button", ui: <Button variant="primary">Click me</Button> },
		{
			name: "button as link",
			ui: (
				<Button asChild variant="primary">
					<a href="...">link</a>
				</Button>
			),
		},
	])("$name", ({ ui }) => {
		const { asFragment } = render(ui)

		expect(asFragment()).toMatchSnapshot()
	})
})
