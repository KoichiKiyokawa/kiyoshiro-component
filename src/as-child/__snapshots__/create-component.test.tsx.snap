// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`create button component with decoration props(outer, inner, ...) > 'button as link' 1`] = `
<DocumentFragment>
  <div
    class="outer"
  >
    <a
      class="button"
      data-variant="primary"
      href="..."
    >
      <div
        class="inner"
      >
        <span
          class="pre1"
        >
          pre1
        </span>
        <span
          class="pre2"
        >
          pre2
        </span>
        link
        <span
          class="post1"
        >
          post1
        </span>
        <span
          class="post2"
        >
          post2
        </span>
      </div>
    </a>
  </div>
</DocumentFragment>
`;

exports[`create button component with decoration props(outer, inner, ...) > 'primary button' 1`] = `
<DocumentFragment>
  <div
    class="outer"
  >
    <button
      class="button"
      data-variant="primary"
      type="button"
    >
      <div
        class="inner"
      >
        <span
          class="pre1"
        >
          pre1
        </span>
        <span
          class="pre2"
        >
          pre2
        </span>
        Click me
        <span
          class="post1"
        >
          post1
        </span>
        <span
          class="post2"
        >
          post2
        </span>
      </div>
    </button>
  </div>
</DocumentFragment>
`;

exports[`create normal button component > 'button as disabled link' 1`] = `
<DocumentFragment>
  <div
    aria-disabled="true"
    class="button"
    data-variant="primary"
  >
    link
  </div>
</DocumentFragment>
`;

exports[`create normal button component > 'button as link' 1`] = `
<DocumentFragment>
  <a
    class="button"
    data-variant="primary"
    href="..."
  >
    link
  </a>
</DocumentFragment>
`;

exports[`create normal button component > 'disabled button' 1`] = `
<DocumentFragment>
  <button
    aria-disabled="true"
    class="button"
    data-variant="primary"
    disabled=""
    type="button"
  >
    Click me
  </button>
</DocumentFragment>
`;

exports[`create normal button component > 'primary button' 1`] = `
<DocumentFragment>
  <button
    class="button"
    data-variant="primary"
    type="button"
  >
    Click me
  </button>
</DocumentFragment>
`;

exports[`create normal button component > 'secondary button' 1`] = `
<DocumentFragment>
  <button
    class="button"
    data-variant="secondary"
    type="button"
  >
    Click me
  </button>
</DocumentFragment>
`;
